subgraphs/_reference_/src/sdk/protocols/perpfutures/position.ts

```diff
@@ -0,0 +1,239 @@
+    this.position.collateralBalanceUSD = constants.BIGDECIMAL_ZERO;
+
+    this.position.collateralInCount = 0;
+    this.position.collateralOutCount = 0;
+    this.position.liquidationCount = 0;
+
+    this.save();
+
+    this.account.openPosition(this.position.side);
+    this.pool.openPosition(this.position.side);
+  }
+
+  /**
+   * Sets the position's fundingrateOpen value.
+   * @param amount
+   */
+  setFundingrateOpen(amount: BigDecimal): void {
+    if (!this.position) return;
+
+    this.position.fundingrateOpen = amount;
+    this.save();
+  }
+
+  /**
+   * Sets the position's fundingrateClosed value.
+   * @param amount
+   */
+  setFundingrateClosed(amount: BigDecimal): void {
+    if (!this.position) return;
+
+    this.position.fundingrateClosed = amount;
+    this.save();
+  }
+
+  /**
+   * Sets the position's leverage value.
+   * @param amount
+   */
+  setLeverage(amount: BigDecimal): void {
+    if (!this.position) return;
+
+    this.position.leverage = amount;
+    this.save();
+  }
+
+  /**
+   * Sets the position's balance value.
+   * @param token
+   * @param amount
+   */
+  setBalance(token: Address, amount: BigInt): void {
+    if (!this.position) return;
+
+    this.position.balance = amount;
+    this.position.balanceUSD = this.protocol
+      .getTokenPricer()
+      .getAmountValueUSD(this.tokens.getOrCreateToken(token), amount);
+    this.save();
+  }
+
+  /**
+   * Sets the position's collateralBalance value.
+   * @param collateralToken
+   * @param amount
+   */
+  setcollateralBalance(collateralToken: Address, amount: BigInt): void {
+    if (!this.position) return;
+
+    this.position.collateralBalance = amount;
+    this.position.collateralBalanceUSD = this.protocol
+      .getTokenPricer()
+      .getAmountValueUSD(this.tokens.getOrCreateToken(collateralToken), amount);
+    this.save();
+  }
+
+  /**
+   * Adds 1 to the account position collateralIn count.
+   */
+  addCollateralIn(): void {
+    if (!this.position) return;
+
+    this.position.collateralInCount += 1;
+    this.save();
+  }
+
+  /**
+   * Adds 1 to the account position collateralOut count.
+   */
+  addCollateralOut(): void {
+    if (!this.position) return;
+
+    this.position.collateralOutCount += 1;
+    this.save();
+  }
+
+  /**
+   * Adds 1 to the account position liquidation count.
+   */
+  addLiquidation(): void {
+    if (!this.position) return;
+
+    this.position.liquidationCount += 1;
+    this.save();
+  }
+
+  closePosition(): void {
+    if (!this.position) return;
+
+    const event = this.protocol.getCurrentEvent();
+    this.position.hashClosed = event.transaction.hash;
+    this.position.blockNumberClosed = event.block.number;
+    this.position.timestampClosed = event.block.timestamp;
+    this.save();
+
+    if (this.account) this.account.closePosition(this.position.side);
+    if (this.pool) this.pool.closePosition(this.position.side);
+  }
+
+  takePositionSnapshot(): void {
```

Comment:
Do we want this to be called by the caller manually? Or can we abstract them from doing so?

Reply:
We can abstract this easily by adding `takeSnapshot` function call at the end of every method in the position class, but I think that's not a good practice, saving an entity many times in a single block can sometimes lead to unknown issues. So I think it's better for a user to call it manually. The same goes for protocol and position snapshots. What are your and @dmelotik thoughts on this?

Reply:
>  saving an entity many times in a single block can sometimes lead to unknown issues

I think that as long as we are always saving the same loaded entity (and not two or more copies loaded at different times) it should be fine. And since we are streamlining the load, update, and update of snapshots of all entities through the SDK I think it is less likely to have the same entity being loaded in multiple places when handling the same event.

We could potentially enforce this at the SDK level. Make the `Protocol` a singleton, so it can only be loaded once, or make it panic if attempted to load multiple times. Same for Pools and accounts. But we can leave that for a future PR if it actually is a problem

Reply:
I don't think we should call `takePositionSnapshot()` on every single function in the position class. I think we only want to take a snapshot when position.balance changes.

I am for what Jaime said though. I prefer making the dev worry about less function calls.